### PR TITLE
Add a new FuzzyPP component.

### PR DESCRIPTION
**Pull request by GlassAndOneHalf**
This commit adds a new "fuzzy preprocessor" which allows users to preprocess their source files, preserving comments, before passing them through to the `fuzzyc2cpg` system.

### CODE REVIEW

**Comment by GlassAndOneHalf** [comment]
All tests are passing locally - I just need to fix the submodules w/ Travis.

**Comment by ltcmelo** [comment]
@GlassAndOneHalf, I'm glad to review this PR. But before doing so, I'd like to understand whether we really need to maitain our own "C preprocessor project" - instead of invoking the user's PP.

Based on a brief Slack chat, you mention the matters of (1) keeping comments and (2) missing headers:

(1) Typically, preprocessors allow you to do that. In Clang/GCC's, this is done with flag `-CC`.
(2) Indeed a problem, but, for certain cases (I believe ours), it should be enough to stub the `.h` in question with an empty file.

**Comment by fabsx00** [comment]
@ltcmelo please trust that we have thought this over and just review the code.

**Comment by AlexDenisov** [comment]
@GlassAndOneHalf currently, you are adding `cxxopts` as a submodule, but committing things like `catch2` into the repo. Would it make sense to also move those third-party things into submodules? This will simplify code reviews, otherwise, it is quite hard to follow.

**Comment by ltcmelo** [comment]
OK @fabsx00. I'll wait until the the "reorg" of the code base, as @AlexDenisov mentions. But, instead of submodules, my suggestion would be use the libs directly.

**Comment by GlassAndOneHalf** [comment]
@AlexDenisov @ltcmelo I've moved `catch2` & `cxxopts` into submodules. `simplecpp` is included directly as I have modified the source slightly in order to preserve comments. You can see the README in the `external/simplecpp` directory to see the change that I made to the source for that project.

**Comment by AlexDenisov** [comment]
@GlassAndOneHalf do you think it worth the effort to use a fork of `simplecpp` with your modifications and also use it as a submodule? I think it might be a cleaner solution.

File: fuzzypp/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                        -Wmissing-include-dirs
+                        -Wno-unused-parameter
+                        -Wuninitialized
+                        -ansi
+                        -pedantic)
+endif()
+
+# Building with Clang is not currently supported.
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    add_compile_options(-Weverything
```
Comment (AlexDenisov) [inline fuzzypp/CMakeLists.txt]: I think this is a bit too much to handle. IIRC some warnings there even contradict the others.

File: fuzzypp/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Dependencies
+---
+- [cxxopts](https://github.com/jarro2783/cxxopts)
+- [simplecpp](https://github.com/danmar/simplecpp)
+- [Catch2](https://github.com/catchorg/Catch2)
+
+FAQ
+---
+Coming soon!
```
Comment (AlexDenisov) [inline fuzzypp/README.md]: Nice readme!

File: fuzzypp/src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Build static library for use by CLI & tests.
+add_library(${FUZZYPP_LIB_NAME} STATIC
+            cliopts.cpp
+            preprocessor.cpp
+            "${CMAKE_SOURCE_DIR}/external/simplecpp/simplecpp.cpp")
+
+target_include_directories(${FUZZYPP_LIB_NAME} PUBLIC
+                           "${CMAKE_SOURCE_DIR}/external/cxxopts/include"
+                           "${CMAKE_SOURCE_DIR}/external/simplecpp")
```
Comment (AlexDenisov) [inline fuzzypp/src/CMakeLists.txt]: Just a comment, no action needed.

There is an option (`SYSTEM`) to mark header search paths as system headers, this way we won't get any warnings that are coming from third-party libraries. Whether it is desired or not is a separate question though 😄

File: fuzzypp/src/cliopts.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace fuzzypp::cliopts {
+    class CliOptions {
+        public:
+            const std::vector<std::string> files;
+            const std::vector<std::string> include_files;
+            const std::vector<std::string> include_paths;
+            const std::vector<std::string> defines;
+            const std::vector<std::string> undefines;
+            const std::string output_directory;
```
Comment (AlexDenisov) [inline fuzzypp/src/cliopts.hpp]: I assume this is required by the `cxxopts` library?

File: fuzzypp/test/cliopts_test.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        std::vector<std::string> { other_header.parent_path() },
+        std::vector<std::string> { "DEF" },
+        std::vector<std::string> { "UNDEF" },
+        output_path
+    };
+
+    auto result = opts.validate_options();
+    REQUIRE(result.has_value());
+    REQUIRE(*result == "Include paths must not contain '..'.");
+}
```
Comment (AlexDenisov) [inline fuzzypp/test/cliopts_test.cpp]: Hats off, very nice test suite 👍

File: fuzzypp/test/main.cpp [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#define CATCH_CONFIG_MAIN
+#include <catch.hpp>
+
+TEST_CASE("1: Test entrypoint", "main") {}
```
Comment (AlexDenisov) [inline fuzzypp/test/main.cpp]: I've never used catch before, so I'm curious what this test does?

Comment (GlassAndOneHalf) [reply fuzzypp/test/main.cpp]: Catch will automatically generate a `main` method for us (see L1). I believe we have to add a dummy test here to get everything to work for reasons unknown. :smile:

File: fuzzypp/src/cliopts.hpp (hunk 4 above)
Comment (GlassAndOneHalf) [reply fuzzypp/src/cliopts.hpp]: That's right, I parse the options using `cxxopts` and put the results into this data structure.

File: fuzzypp/CMakeLists.txt (hunk 1 above)
Comment (GlassAndOneHalf) [reply fuzzypp/CMakeLists.txt]: I need to play around with the Clang support. I don't think it compiles at the moment (not even due to having `-Weverything` turned on!)

**Comment by GlassAndOneHalf** [comment]
Going to hold off merging this until I can get a SL fork of simplecpp.

File: fuzzypp/test/main.cpp (hunk 6 above)
Comment (AlexDenisov) [reply fuzzypp/test/main.cpp]: 👍

File: fuzzypp/src/preprocessor.hpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#define FUZZYPP_PREPROCESSOR_HPP
+
+#include <string>
+
+#include <simplecpp.h>
+
+#include "cliopts.hpp"
+
+namespace fuzzypp::preprocessor {
+    class FuzzyPreprocessor {
```
Comment (ltcmelo) [inline fuzzypp/src/preprocessor.hpp]: Silly note, but the preprocessor itself isn't fuzzy ;). This name can be misleading, perhaps just `Preprocessor` or `Fuzzyc2cpgPreprocessor`, or `Fuzzyc2cpgPP`?

File: fuzzypp/src/preprocessor.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        std::for_each(options.files.cbegin(),
+                      options.files.cend(),
+                      write_to_file);
+    }
+
+    const simplecpp::DUI
+    FuzzyPreprocessor::generate_simplecpp_opts(const fuzzypp::cliopts::CliOptions& options) {
+        simplecpp::DUI simple_opts;
+
+        simple_opts.includes = std::list<std::string> { options.include_files.cbegin(), options.include_files.cend() };
```
Comment (ltcmelo) [inline fuzzypp/src/preprocessor.cpp]: Why not making the containers in `CliOptions` an `std::list` too, so that you can copy (or even move) them here?

File: fuzzypp/test/preprocessor_test.cpp [hunk 9]
```diff
@@ -0,0 +1,332 @@
+#include <catch.hpp>
```
Comment (ltcmelo) [inline fuzzypp/test/preprocessor_test.cpp]: General question concerning behaviour and tests: what's our expecation for macros? For instance, would fuzzyc2cpg interpret function- and object-like macros as is? If they must be expanded, I'd assume that **simplecpp** itself has tests for those, right? Does it handle [self referential macros](https://gcc.gnu.org/onlinedocs/gcc-3.3.6/cpp/Self_002dReferential-Macros.html) well?

File: fuzzypp/src/preprocessor.cpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        simple_opts.includePaths = std::list<std::string> { options.include_paths.cbegin(), options.include_paths.cend() };
+        simple_opts.defines = std::list<std::string> { options.defines.cbegin(), options.defines.cend() };
+        simple_opts.undefined = std::set<std::string> { options.undefines.cbegin(), options.undefines.cend() };
```
Comment (ltcmelo) [inline fuzzypp/src/preprocessor.cpp]: `undefined` x `undefines`

File: fuzzypp/external/simplecpp/README.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
+This is a version of [simplecpp](https://github.com/danmar/simplecpp), modified to retain comments in the preprocessor output.
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/external/simplecpp/README.md]: Yeah I think this would be the idea long term plan - I just wanted to make a quick & dirty change here :wink:

File: fuzzypp/src/cliopts.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+                    output
+                }
+            };
+        } catch (const cxxopts::OptionException& e) {
+            std::cerr << "Error occured whilst parsing command line arguments: " << e.what() << std::endl;
+            return std::optional<CliOptions> {};
+        }
+    }
+
+    inline const std::vector<std::string>
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/src/cliopts.cpp]: Are you sure that's correct? I can't find anything stating this.

https://en.cppreference.com/w/cpp/language/inline
https://isocpp.org/wiki/faq/inline-functions#where-to-put-inline-keyword

File: fuzzypp/src/cliopts.cpp [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        if (std::filesystem::exists(output_directory) && !std::filesystem::is_directory(output_directory)) {
+            return std::optional<std::string> { "The specified output directory must be a directory." };
+        }
+
+        for (const auto& file : files) {
+            std::filesystem::path p { file };
+            if (!is_path_valid(p)) {
+                return std::optional<std::string> { "File paths must not contain '..'." };
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/src/cliopts.cpp]: When we generate the output, we try to mimic the existing structure of the source code in the filesystem. If the user passes the program a file path with `..`, it can play havoc with this.

File: fuzzypp/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,55 @@
+cmake_minimum_required(VERSION 3.10)
+
+project(fuzzyppcli VERSION 0.0.1 LANGUAGES CXX)
+
+# Set compile options. 
+add_compile_options(-O3)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+    add_compile_options(-Wall
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/CMakeLists.txt]: This generates a _lot_ of errors from `cxxopts`. Will leave this out for the time being.

File: fuzzypp/src/cliopts.hpp [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+                       files(_files), include_files(_include_files), 
+                       include_paths(_include_paths), defines(_defines), 
+                       undefines(_undefines), output_directory(_output_directory) {}
+
+            CliOptions() = delete;
+            CliOptions(CliOptions&) = delete;
+            CliOptions(CliOptions&&) = default;
+
+            static const std::optional<CliOptions> 
+            parse_command_line(int argc, char* argv[]);
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/src/cliopts.hpp]: We produce a library from the source, so clients are able to construct their own `CliOptions` without having to fake cmd line args.

File: fuzzypp/src/preprocessor.cpp [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+#include "preprocessor.hpp"
+
+namespace fuzzypp::preprocessor {
+    void 
+    FuzzyPreprocessor::preprocess(const fuzzypp::cliopts::CliOptions& options) {
+        auto opts = generate_simplecpp_opts(options);
+        const auto output_path = std::filesystem::path { options.output_directory };
+
+        auto write_to_file = [&](const auto& filename) {
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/src/preprocessor.cpp]: Nice catch!

File: fuzzypp/src/preprocessor.cpp [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+                      options.files.cend(),
+                      write_to_file);
+    }
+
+    const simplecpp::DUI
+    FuzzyPreprocessor::generate_simplecpp_opts(const fuzzypp::cliopts::CliOptions& options) {
+        simplecpp::DUI simple_opts;
+
+        simple_opts.includes = std::list<std::string> { options.include_files.cbegin(), options.include_files.cend() };
+        simple_opts.includePaths = std::list<std::string> { options.include_paths.cbegin(), options.include_paths.cend() };
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/src/preprocessor.cpp]: The left hand side of the expression is from a third party library :(

File: fuzzypp/src/preprocessor.cpp [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+        return simple_opts;
+    }
+
+    const std::string
+    FuzzyPreprocessor::stringify(const std::string& filename, const simplecpp::DUI& options) {
+        simplecpp::OutputList output_list;
+        std::vector<std::string> files;
+        
+        std::ifstream input_file { filename };
```
[the remark this answers is not in the document]
Comment (GlassAndOneHalf) [reply fuzzypp/src/preprocessor.cpp]: We check them when performing validation on the options in `cliopts.cpp`. Perhaps it would be a good idea to do this anyway? What do you think?

File: fuzzypp/test/preprocessor_test.cpp (hunk 9 above)
Comment (GlassAndOneHalf) [reply fuzzypp/test/preprocessor_test.cpp]: Yes, I believe it already handles these cases. It won't stop you from infinitely recursing, though :wink:

File: fuzzypp/src/preprocessor.cpp (hunk 10 above)
Comment (GlassAndOneHalf) [reply fuzzypp/src/preprocessor.cpp]: As with the `includePaths` - stuck with this name :(

File: fuzzypp/src/preprocessor.cpp (hunk 8 above)
Comment (GlassAndOneHalf) [reply fuzzypp/src/preprocessor.cpp]: Just tried this out. Unfortunately it seems `cxxopts` doesn't support parsing cmd args as a list :(

File: fuzzypp/src/cliopts.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (ltcmelo) [reply fuzzypp/src/cliopts.cpp]: Yes, how would the compiler _inline_ the function if it can't see it? The compiler can't look across different .cpp, only in .h.  Note that putting an inline function **outside the class body** doesn't mean that it has to be in the .cpp, you can (and should) have those definitions in the .h. Check that same page you linked (see screenshot):

Screen Shot 2019-10-11 at 11 09 45

File: fuzzypp/src/preprocessor.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ltcmelo) [reply fuzzypp/src/preprocessor.cpp]: Generally yes. Even if the path is valid, you'll only know whether you can actually open the file through the stream.
